Copy systests imagePullSecrets from default-ns to target-ns

elakito opened the pull request:
Signed-off-by: Aki Yoshida <elakito@gmail.com>

### Type of change
- Enhancement or Bug fix

### Description
Although the documentation (https://github.com/strimzi/strimzi-kafka-operator/blob/main/development-docs/TESTING.md#using-private-registries) mentions that systemtests can be executed against a protected registry by setting the image pull secret name to SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET, the images are not pulled.

The background to this issue is found at https://cloud-native.slack.com/archives/CMH3Q3SNP/p1624464225264000

In this proposed change, the provided secret placed in the default namespace is copied to the target namespace and the installed resources will refer to this secret so that the images can be pulled.

I would appreciate for any comments. It was not clear to me which approach is preferable, whether to refer to the secret over the pod template's imagePullSecrets or over env variable STRIMZI_IMAGE_PULL_SECRETS. I have not added any tests yet but tested with the existing systemtests against a protected registry. If the approach proposed is correct, I can add some tests.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

scholzj commented:
/azp run regression

Frawless commented on systemtest/src/main/java/io/strimzi/systemtest/resources/operator/BundleResource.java:
```diff
@@ -196,6 +200,11 @@ public DeploymentBuilder buildBundleDeployment() {
         // Apply updated env variables
         clusterOperator.getSpec().getTemplate().getSpec().getContainers().get(0).setEnv(envVars);
 
+        if (SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET != null && !SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET.isEmpty()) {
```
```suggestion
        if (Environment.SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET != null && !Environment.SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET.isEmpty()) {
```

Could you please use it as it's shown in the suggestion? We use variables from Environment this way across the whole project.

elakito replied:
Yes. I'll change it.

elakito replied:
rebased to resolve the conflicts and changed the variable references as requested

Frawless commented on systemtest/src/main/java/io/strimzi/systemtest/utils/StUtils.java:
```diff
@@ -373,4 +377,26 @@ public static boolean isParallelNamespaceTest(ExtensionContext extensionContext)
+        if (kubeClient("default").getSecret(SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET) == null) {
+            throw new RuntimeException(SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET + " is not in the default namespace!");
+        }
+        Secret pullSecret = kubeClient("default").getSecret(SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET);
```
Is the namespace name the same for openshift and kubernetes?

elakito replied:
I am not sure if I understand your question. Is the default namespace "default" not suitable for openshift?

Frawless replied:
It will work. My question was if kubernetes and openshift will always contain `default` namespace and according to some docs, it should so I think we are fine with this.

Frawless commented:
/azp run regression

elakito commented:
I see LoggingChangeST failed in the above regression systemtests, but it looks like these errors are unrelated to the change and just happen sporadically? I could run those tests successfully.

Frawless commented:
@elakito I restarted the job. I am pretty sure it's not related to your changes.

elakito commented:
@Frawless I think the job was not restarted. It shows the rest from two days ago (20210706). Could you restart it again?

Frawless commented:
@elakito  it was restarted based on timestamp. @im-konge  is trying to find out why it's failing

elakito commented:
@Frawless Thanks. I see you restarted that particular regression bundle test. So, I didn't see Azure bot's message in this ticket and thought it was not restarted. I'll also look at it again.

Frawless commented:
@elakito  it seems there is some race condition. @im-konge will provide more info later today

Frawless commented:
We agreed it's some race condition connected to azure. We will investigate it, but we don't think it should block merging this PR.

scholzj commented on systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaTemplates.java:
```diff
@@ -227,6 +236,10 @@ private static KafkaBuilder defaultKafka(Kafka kafka, String name, int kafkaRepl
                     .endTopicOperator()
                 .endEntityOperator()
             .endSpec();
+        if (Environment.SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET != null && !Environment.SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET.isEmpty()) {
```
Maybe a stupid question ... but why to use this and not configure the secret in the operator deployment directly? (using the `STRIMZI_IMAGE_PULL_SECRETS` env var which can configure secrets which will be passed to the operands?) It would seem cleaner than adding it to every custom resource. It would be IMHO also more stable because nobody would overwrite this in the ST itself by mistake.

elakito replied:
@scholzj I thought about passing the secret name with env using ContainerTemplate but I didn't think about passing the variable directly without manipulating the in-memory custom resource. Is that what you are suggesting? That sounds good.

Frawless replied:
To CO deployment you can insert it to container env part. It will work for all Strimzi components except our clients, which are not handled by CO.

elakito replied:
That means, we modify the in-memory resource with ContainerTemplate's env part to add this variable. It was not clear to me if there was a way to pass the variable without modifying the in-memory resource.

elakito replied:
Thanks for the comments. I updated the code to use the env based image-pull-secret injection for kafka.

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaTemplates.java:
```diff
@@ -246,6 +259,28 @@ public static KafkaBuilder kafkaFromYaml(File yamlFile, String clusterName, int
+        ZookeeperClusterTemplate zookeeperClusterTemplate = new ZookeeperClusterTemplateBuilder().withPod(podTemplate).build();
+        EntityOperatorTemplate entityOperatorTemplate = new EntityOperatorTemplateBuilder().withPod(podTemplate).build();
+
+        builder.editSpec()
+            .editKafka()
+                .withTemplate(kafkaClusterTemplate)
+            .endKafka()
+            .editZookeeper()
+                .withTemplate(zookeeperClusterTemplate)
+            .endZookeeper()
+            .editEntityOperator()
+                .withTemplate(entityOperatorTemplate)
+            .endEntityOperator()
+            .endSpec();
```
```suggestion
        builder
            .editSpec()
                .editKafka()
                    .withTemplate(kafkaClusterTemplate)
                .endKafka()
                .editZookeeper()
                    .withTemplate(zookeeperClusterTemplate)
                .endZookeeper()
                .editEntityOperator()
                    .withTemplate(entityOperatorTemplate)
                .endEntityOperator()
            .endSpec();
```
Just indent again

elakito replied:
For kafka, I switched to use the env variable based image-pull-secret injection and removed the template editing code.

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/utils/StUtils.java:
```diff
@@ -383,4 +386,25 @@ public static boolean isParallelNamespaceTest(ExtensionContext extensionContext)
+            throw new RuntimeException(Environment.SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET + " is not in the default namespace!");
+        }
+        Secret pullSecret = kubeClient("default").getSecret(Environment.SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET);
+        kubeClient(namespace).createSecret(new SecretBuilder()
+                .withApiVersion("v1")
+                .withKind("Secret")
+                .withNewMetadata()
+                .withName(Environment.SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET)
+                .endMetadata()
+                .withType("kubernetes.io/dockerconfigjson")
+                .withData(Collections.singletonMap(".dockerconfigjson", pullSecret.getData().get(".dockerconfigjson")))
+                .build());
+    }
```
```suggestion
        kubeClient(namespace).createSecret(new SecretBuilder()
                .withApiVersion("v1")
                .withKind("Secret")
                .withNewMetadata()
                    .withName(Environment.SYSTEM_TEST_STRIMZI_IMAGE_PULL_SECRET)
                .endMetadata()
                .withType("kubernetes.io/dockerconfigjson")
                .withData(Collections.singletonMap(".dockerconfigjson", pullSecret.getData().get(".dockerconfigjson")))
                .build());
    }
```
Can you please add indentation?

elakito replied:
I fixed the indentation.

im-konge reviewed (approved):
Just indentation comments, anyway LGTM. Thanks for the PR :) and how @Frawless mentioned above, test failure is some kind of race condition, which doesn't block this to be merged.

im-konge commented:
Hey @elakito , my fix for the failing test was merged! (#5278) So now, if you rebase it, AZP should pass.

elakito commented:
I switched to use the env-based secret injection for kafka and verified it is working.  I also rebased the changes on the current main which has @im-konge 's update. Regarding the test possibility, raised by @see-quick, I think some unit tests may be more suitable than creating a test against a protected registry because that would require a complex setup.

Frawless commented:
/azp run regression

scholzj commented:
So, should we merge this? @Frawless?

Frawless reviewed (approved):
Yes we can

scholzj commented:
@elakito Thanks for the PR!
